src/cache/intern.rs

```diff
@@ -0,0 +1,78 @@
+use cfg_if::cfg_if;
+
+
+cfg_if! {
+    if #[cfg(feature = "enable-interning")] {
+        use std::thread_local;
+        use std::string::String;
+        use std::borrow::ToOwned;
+        use std::cell::RefCell;
+        use std::collections::HashMap;
+        use crate::JsValue;
+        use crate::convert::IntoWasmAbi;
+
+        struct Cache {
+            entries: RefCell<HashMap<String, JsValue>>,
+        }
+
+        thread_local! {
+            static CACHE: Cache = Cache {
+                entries: RefCell::new(HashMap::new()),
+            };
+        }
+
+        /// This returns the raw index of the cached JsValue, so you must take care
+        /// so that you don't use it after it is freed.
+        pub(crate) fn unsafe_get_str(s: &str) -> Option<<JsValue as IntoWasmAbi>::Abi> {
+            CACHE.with(|cache| {
+                let cache = cache.entries.borrow();
+
+                cache.get(s).map(|x| x.into_abi())
+            })
+        }
+
+        fn intern_str(key: &str) {
+            CACHE.with(|cache| {
+                let mut cache = cache.entries.borrow_mut();
+
+                // Can't use `entry` because `entry` requires a `String`
+                if !cache.contains_key(key) {
+                    cache.insert(key.to_owned(), JsValue::from(key));
+                }
+            })
+        }
+    }
+}
+
+
+/// Interns Rust strings so that it's much faster to send them to JS.
+///
+/// Sending strings from Rust to JS is slow, because it has to do a full `O(n)`
+/// copy and *also* encode from UTF-8 to UTF-16. This must be done every single
+/// time a string is sent to JS.
+///
+/// If you are sending the same string multiple times, you can call this `intern`
+/// function, which simply returns its argument unchanged:
+///
+/// ```rust
+/// # use wasm_bindgen::intern;
+/// intern("foo") // returns "foo"
+/// ```
+///
+/// However, if you enable the `"enable-interning"` feature for wasm-bindgen,
+/// then it will add the string into an internal cache.
+///
+/// When you send that cached string to JS, it will look it up in the cache,
+/// which completely avoids the `O(n)` copy and encoding. This has a significant
+/// speed boost (as high as 783%)!
+///
+/// However, there is a small cost to this caching, so you shouldn't cache every
+/// string. Only cache strings which have a high likelihood of being sent
+/// to JS multiple times.
+#[inline]
+pub fn intern(s: &str) -> &str {
```

Comment:
Perhaps this could be renamed to `insert_in_intern_cache` and not have a return value? (or something like that name)

I would expect an `intern` function to return some sort of different type which is zero cost to pass around (aka `JsValue::from(&str)`) so here it's more of "populate the cache so the next time it can be faster". 

It may also be worth pointing out that using `intern` is always going to be slower than just using `JsValue::from` because we still have to do some sort of lookup in the intern cache.

Reply:
It might also be good to throw "maybe" in the name somewhere since if the feature isn't turned on this doesn't actually do anything.

Reply:
> I would expect an intern function to return some sort of different type which is zero cost to pass around (aka JsValue::from(&str)) so here it's more of "populate the cache so the next time it can be faster".

Yes, that's very intuitive, but that's not actually possible. It wouldn't be the first wasm-bindgen API which is a bit unintuitive for technical reasons.

> Perhaps this could be renamed to insert_in_intern_cache and not have a return value? (or something like that name)

I had considered that, but that's a *really* bad UX. The choice is between these two options:

```rust
document.create_text_node(intern("foo"));
```

```rust
let s = "foo";
insert_in_intern_cache(s);
document.create_text_node(s);
```

And it gets even worse when you consider things like dominator, which provide a convenient high-level API on top of the DOM:

```rust
html!("div", {
    .style("foo", intern("bar"))
})
```

```rust
html!("div", {
    .style("foo", {
        let s = "bar";
        insert_into_intern_cache(s);
        s
    })
})
```

Basically, every time you want to intern a string you *also* want to immediately use it.

The current API is designed for that, rather than the *incredibly* rare case of "I want to intern this string but *not* immediately use it".

> It might also be good to throw "maybe" in the name somewhere since if the feature isn't turned on this doesn't actually do anything.

I don't think that's necessary. I view it as being a performance hint, it's not guaranteed that the strings will be cached anyways (e.g. they might get evicted if the cache grows too large).

Reply:
Yeah I definitely agree there's different designs here, and I don't really mind one way or another. Looks like you feel stronger about this than I, so leaving it as `intern` seems fine.